Source/NSDateInterval.m

```diff
@@ -0,0 +1,214 @@
+
+// Init
+- (instancetype)init
+{
+  self = [super init];
+  if(self != nil)
+    {
+      _startDate = nil;
+      _duration = 0.0;
+    }
+  return self;
+}
+
+- (instancetype)initWithStartDate:(NSDate *)startDate 
+                         duration:(NSTimeInterval)duration
+{
+  self = [super init];
+  if(self != nil)
+    {
+      ASSIGNCOPY(_startDate, startDate);
+      _duration = duration;
+    }
+  return self;
+}
+
+- (instancetype)initWithStartDate:(NSDate *)startDate 
+                          endDate:(NSDate *)endDate
+{
+  self = [super init];
+  if(self != nil)
+    {
+      ASSIGNCOPY(_startDate, startDate);
+      ASSIGNCOPY(_endDate, endDate);
+      _duration = [endDate timeIntervalSinceDate: startDate];
+    }
+  return self;  
+}
+
+- (instancetype) initWithCoder: (NSCoder *)coder
+{
+  return nil;
+}
+
+- (void) encodeWithCoder: (NSCoder *)coder
+{
+}
+
+- (id) copyWithZone: (NSZone *)zone
+{
+  return [[[self class] allocWithZone: zone]
+           initWithStartDate: _startDate
+                    duration: _duration];
+}
+
+- (void) dealloc
+{
+  RELEASE(_startDate);
+  [super dealloc];
+}
+
+// Access
+- (NSDate *) startDate
+{
+  return _startDate;
+}
+
+- (void) setStartDate: (NSDate *)startDate
+{
+  ASSIGNCOPY(_startDate, startDate);
+}
+
+- (NSDate *) endDate
+{
+  return [_startDate dateByAddingTimeInterval: _duration];
+}
+
+- (void) setEndDate: (NSDate *)endDate
+{
+  _duration = [endDate timeIntervalSinceDate: _startDate];
+}
+
+- (NSTimeInterval) duration
+{
+  return _duration;
+}
+
+- (void) setDuration: (NSTimeInterval)duration
+{
+  _duration = duration;
+}
+
+// Compare
+- (NSComparisonResult) compare: (NSDateInterval *)dateInterval
+{
+  // NSOrderedAscending
+  if([_startDate isEqualToDate: [dateInterval startDate]] &&
+     _duration < [dateInterval duration])
+    return NSOrderedAscending;
+  
+  if([_startDate compare: [dateInterval startDate]] == NSOrderedAscending)
+    return NSOrderedAscending;
+
+  // NSOrderedSame 
+  if([self isEqualToDateInterval: dateInterval])
+    return NSOrderedSame;
+
+  // NSOrderedDescending
+  if([_startDate isEqualToDate: [dateInterval startDate]] &&
+     _duration > [dateInterval duration])
+    return NSOrderedDescending;
+  
+  if([_startDate compare: [dateInterval startDate]] == NSOrderedDescending)
+    return NSOrderedDescending;
+
+  return 0;
+}
+
+- (BOOL) isEqualToDateInterval: (NSDateInterval *)dateInterval
+{
+  return ([_startDate isEqualToDate: [dateInterval startDate]] &&
+          _duration == [dateInterval duration]);
+}
+
+// Determine
+- (BOOL) intersectsDateInterval: (NSDateInterval *)dateInterval
+{
+  return [self intersectionWithDateInterval: dateInterval] != nil;
+}
+
+- (NSDateInterval *) intersectionWithDateInterval: (NSDateInterval *)dateInterval
+{
+  NSDateInterval *result = nil;
+  NSArray *array = [NSArray arrayWithObjects: self, dateInterval, nil];
+  NSArray *sortedArray = [array sortedArrayUsingSelector: @selector(compare:)];
```

Comment:
The sorting here is completely useless.

Reply:
If the one with the earliest start date is first it makes the subsequent calculation easier.

Reply:
No, it doesn't. Especially not the way you wrote the subsequent code.

Reply:
Removed